File: hooks/checks/feature-isolation/shared_usage.py
import ast
import sys
from collections import defaultdict
from pathlib import Path
from typing import NamedTuple

SHARED_PACKAGE = "shared"
FEATURES_PACKAGE = "features"
MINIMUM_FEATURES = 2


class LonelyModule(NamedTuple):
    path: str
    features: tuple[str, ...]


class ModuleName:
    def of(self, path: Path) -> str:
        parts = path.parts
        index = parts.index("src")
        trail = parts[index + 1 :]
        dotted = ".".join(trail).removesuffix(".py")

        return dotted.removesuffix(".__init__")


class OwningFeature:
    def of(self, path: Path) -> str | None:
        parts = path.parts

        if FEATURES_PACKAGE not in parts:
            return None

        index = parts.index(FEATURES_PACKAGE)

        if index + 1 >= len(parts):
            return None

        return parts[index + 1]


class SharedImports:
    def in_(self, tree: ast.Module) -> set[str]:
        found: set[str] = set()

        for node in ast.walk(tree):
            for module in self._modules_of(node):
                if module.split(".")[0] == SHARED_PACKAGE:
                    found.add(module)

        return found

    def _modules_of(self, node: ast.AST) -> list[str]:
        if isinstance(node, ast.ImportFrom):
            return [] if node.level else [node.module or ""]

        if isinstance(node, ast.Import):
            return [alias.name for alias in node.names]

        return []


class ImportedModule:
    def owning(self, imported: str, known: set[str]) -> str | None:
        candidate = imported

        while candidate:
            if candidate in known:
                return candidate

            candidate = candidate.rpartition(".")[0]

        return None


class LonelyModules:
    def find_in(self, paths: list[str]) -> list[LonelyModule]:
        modules = {ModuleName().of(Path(p)): p for p in paths}
        shared = {
            name for name in modules if name.split(".")[0] == SHARED_PACKAGE
        }
        callers: dict[str, set[str]] = defaultdict(set)
        kept: set[str] = set()

        for path in paths:
            self._record(Path(path), shared, callers, kept)

        return self._lonely(shared, modules, callers, kept)

    def _record(
        self,
        path: Path,
        shared: set[str],
        callers: dict[str, set[str]],
        kept: set[str],
    ) -> None:
        tree = ast.parse(path.read_text(encoding="utf-8"))
        feature = OwningFeature().of(path)
        owning = ImportedModule()

        for imported in SharedImports().in_(tree):
            owner = owning.owning(imported, shared)

            if owner is None:
                continue

            if feature is not None:
                callers[owner].add(feature)
            elif ModuleName().of(path) != owner:
                kept.add(owner)

    def _lonely(
        self,
        shared: set[str],
        modules: dict[str, str],
        callers: dict[str, set[str]],
        kept: set[str],
    ) -> list[LonelyModule]:
        found: list[LonelyModule] = []

        for name in sorted(shared):
            if name in kept:
                continue

            features = callers[name]

            if len(features) >= MINIMUM_FEATURES:
                continue

            found.append(LonelyModule(modules[name], tuple(sorted(features))))

        return found


source_paths = sys.stdin.read().split()

for lonely in LonelyModules().find_in(source_paths):
    if lonely.features:
        reason = f"only {lonely.features[0]} imports it"
    else:
        reason = "no feature imports it"

    _ = sys.stdout.write(f"{lonely.path}: {reason}\n")

File: hooks/checks/feature-isolation/shared_usage.js
const path = require("path");
const {
  importsIn,
  moduleNameOf,
  readPathsFromStandardInput,
  resolveImport,
  sourceRootOf,
} = require("./module_paths");

const SHARED_PACKAGE = "shared";
const FEATURES_DIRECTORY = "features";
const MINIMUM_FEATURES = 2;

const typescript = require(process.argv[2]);

function owningFeatureOf(filePath) {
  const segments = path.resolve(filePath).split(path.sep);
  const index = segments.lastIndexOf(FEATURES_DIRECTORY);

  if (index === -1 || index + 1 >= segments.length) return null;

  return segments[index + 1];
}

function sharedModulesAmong(paths) {
  const modules = new Map();

  for (const filePath of paths) {
    const sourceRoot = sourceRootOf(filePath);

    if (sourceRoot === null) continue;

    const name = moduleNameOf(sourceRoot, filePath);

    if (name.split("/")[0] === SHARED_PACKAGE) modules.set(name, filePath);
  }

  return modules;
}

function sharedImportsIn(filePath, shared) {
  const sourceRoot = sourceRootOf(filePath);
  const found = [];

  if (sourceRoot === null) return found;

  for (const reference of importsIn(typescript, filePath)) {
    const resolved = resolveImport(filePath, reference.specifier);

    if (resolved === null) continue;

    const name = moduleNameOf(sourceRoot, resolved);

    if (shared.has(name)) found.push(name);
  }

  return found;
}

function recordUsage(filePath, shared, callers, kept) {
  const feature = owningFeatureOf(filePath);
  const sourceRoot = sourceRootOf(filePath);
  const ownName =
    sourceRoot === null ? null : moduleNameOf(sourceRoot, filePath);

  for (const name of sharedImportsIn(filePath, shared)) {
    if (feature !== null) {
      callers.get(name).add(feature);
    } else if (ownName !== name) {
      kept.add(name);
    }
  }
}

function lonelyModulesAmong(paths) {
  const shared = sharedModulesAmong(paths);
  const callers = new Map([...shared.keys()].map((name) => [name, new Set()]));
  const kept = new Set();

  for (const filePath of paths) recordUsage(filePath, shared, callers, kept);

  return [...shared.keys()].sort().flatMap((name) => {
    const features = [...callers.get(name)].sort();

    if (kept.has(name) || features.length >= MINIMUM_FEATURES) return [];

    return [reportFor(shared.get(name), features)];
  });
}

function reportFor(filePath, features) {
  const reason =
    features.length === 0
      ? "no feature imports it"
      : `only ${features[0]} imports it`;

  return `${filePath}: ${reason}`;
}

for (const lonely of lonelyModulesAmong(readPathsFromStandardInput())) {
  process.stdout.write(lonely + "\n");
}
